src/modules/keyboard.rs

```diff
@@ -149,22 +187,47 @@ impl Module<gtk::Box> for KeysModule {
                 match ev {
                     Event::Device => {
                         for key in [Key::Caps, Key::Num, Key::Scroll] {
-                            module_update!(
-                                tx,
-                                KeyEvent {
-                                    key: Key::Caps,
-                                    state: client.get_state(key)
-                                }
-                            );
+                            let event = KeyEvent {
+                                key,
+                                state: client.get_state(key),
+                            };
+                            module_update!(tx, KeyboardUpdate::Key(event));
                         }
                     }
                     Event::Key(ev) => {
-                        send_async!(tx, ModuleUpdateEvent::Update(ev));
+                        module_update!(tx, KeyboardUpdate::Key(ev));
                     }
                 }
             }
         });
 
+        let client = context.ironbar.clients.borrow_mut().keyboard_layout()?;
```

Comment:
As this client takes no parameters, you should be able can use the shorthand here:

```suggestion
        let client = context.client::<compositor::KeyboardLayoutClient>()?;
```

(might need a `dyn` on that, or it might not work because it's a trait I'm not 100%)

Reply:
It's `dyn` + `try_client`